cloud/scope/managedcontrolplane.go

```diff
@@ -0,0 +1,104 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package scope
+
+import (
+	"context"
+
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog/klogr"
+	azexpv1 "sigs.k8s.io/cluster-api-provider-azure/exp/api/v1alpha3"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
+	expv1 "sigs.k8s.io/cluster-api/exp/api/v1alpha3"
+
+	"sigs.k8s.io/cluster-api/util/patch"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// ManagedControlPlaneScopeParams defines the input parameters used to create a new Scope.
+type ManagedControlPlaneScopeParams struct {
+	AzureClients
+	Client           client.Client
+	Logger           logr.Logger
+	Cluster          *clusterv1.Cluster
+	ControlPlane     *azexpv1.AzureManagedControlPlane
+	Context          context.Context
+	InfraMachinePool *azexpv1.AzureManagedMachinePool
+	MachinePool      *expv1.MachinePool
+	PatchTarget      runtime.Object
+}
+
+// NewManagedControlPlaneScope creates a new Scope from the supplied parameters.
+// This is meant to be called for each reconcile iteration.
+func NewManagedControlPlaneScope(ctx context.Context, params ManagedControlPlaneScopeParams) (*ManagedControlPlaneScope, error) {
+	if params.Cluster == nil {
+		return nil, errors.New("failed to generate new scope from nil Cluster")
+	}
+
+	if params.ControlPlane == nil {
+		return nil, errors.New("failed to generate new scope from nil ControlPlane")
+	}
+
+	if params.Logger == nil {
+		params.Logger = klogr.New()
+	}
+
+	err := params.AzureClients.setCredentials()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create Azure session")
+	}
+
+	helper, err := patch.NewHelper(params.PatchTarget, params.Client)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to init patch helper")
+	}
+
+	return &ManagedControlPlaneScope{
+		Context:          ctx,
```

Comment:
Having context baked into the struct makes me uncomfortable. Usually, it should be provided to functions, not constructors.

Reply:
I generally agree and would prefer we align all impls to follow that pattern. Given that scope is effectively a context for the request, I'm fine with it for now and don't think we're doing anything too terrible.